teps/0039-add-variable-retries-and-retrycount.md

```diff
@@ -0,0 +1,108 @@
+creation-date: '2020-10-14'
+last-updated: '2020-12-21'
+authors:
+- '@yaoxiaoqi'
+---
+
+# TEP-0039: Add Variable `retries` and `retry-count`
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [Proposal](#proposal)
+  - [Alternatives](#alternatives)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes adding current `retries` time and total `retry-count` to Pipeline variables in
+order to send notifications after a PipelineTask exhausted its retries.
+
+## Motivation
+
+PipelineTask can be rerun after failure when user specifies `retries` in PipelineTask. In
+that case, exposing the retry-count can give user more information. User can dispatch a
+notification after certain times of failure, or customize logs according to current `retries`
+time.
+
+### Goals
+
+- add 2 variables indicating retries and retry-count of a PipelineTask
+
+### Non-Goals
+
+- execute different PipelineTask based on current `retries` time
+- build a new mechanism for Tekton notification
+
+### Use Cases
+
+`PipelineTask` author wants to send a Slack notification only if it fails after specified
+`retries`, not every time it fails because the user wants less interruption.
+
+## Proposal
+
+Due to the use case is very clear, the variable scope should be as small as possible. Firstly,
+a PipelineTask's `retries` and `retry-count` should not be exposed to other PipelineTask. Secondly,
+these 2 variables should only exist in the context of PipelineRun .
+
+Therefore, these 2 variable substitutions `context.pipelineTask.retries` and `context.pipelineTask.retry-count`
+can be added to support the use case:
+
+1. To retrieve `retries`, we can use `Retries` stored in `PipelineTask` struct
+2. To retrieve `retry-count`, we can use the length of `RetriesStatus` in `TaskRunStatus`
+
+Currently, Tekton does the parameter and context substitution (`ApplyParameters`, `ApplyContext`, etc)
+before determining the next set of TaskRuns to create. However, if the current TaskRun fails, its status
+will be appended to `RetriesStatus` and then cleared when creating next TaskRun. We should redo
+the context substitution at this time to ensure the `retries` variable indicates the current retries time.
+
+Code example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: retry-example
+spec:
+  tasks:
+  - name: retry-me
+    retries: 5
+    taskSpec:
+      steps:
+      - image: alpine:3.12.0
+        script: |
+          #!/usr/bin/env sh
+          if [ "$(context.pipelineTask.retry-count)" == "$(context.pipelineTask.retries)" ]; then
+            echo "This is the last retry. \n"
+            exit 1
+          fi
+      - image: ubuntu
+        script: |
+          #!/usr/bin/env sh
+          /path/to/<program-that-fails>
+```
+
+In this example, Tekton will echo `This is the last retry` when retry count is equal to 5.
```

Comment:
My understanding is this is not possible. `$(context.pipelineTask.retries)` is 5 in this example since its specified in the schema but `$(context.pipelineTask.retry-count)` can not be equal to the number of retries. `$(context.pipelineTask.retry-count)` represents the length of `RetriesStatus` which would be 4 when fifth retry is happening unless you always add one to it 🙃  So the very first execution without any `RetriesStatus` would set `$(context.pipelineTask.retry-count)` to 1?

Reply:
the controller declares a task failure when number of `RetriesStatus` is >= `retries` i.e. controller does not schedule that task once the number of `RetriesStatus` matches `retries`:

https://github.com/tektoncd/pipeline/blob/2e26eef219ddfabf505041867d4825b47da8cb36/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L96-L107

Reply:
It's possible as far as I know. The replacement of contexts means to happen after the update of RetriesStatus. For example, let's say `tr` is a PipelineTask that means to fail, and its Retries is set to 3. We might misunderstand that the controller would run `tr` 3 times in total. But according to our code, `tr` will run 4 times in total (because of this line, it executes before appending RetriesStatus).

https://github.com/tektoncd/pipeline/blob/2e26eef219ddfabf505041867d4825b47da8cb36/pkg/reconciler/pipelinerun/resources/pipelinerunstate.go#L232

let's say the order of taskrun and its retries is `tr#0 -> tr_retry#1 -> tr_retry#2 -> tr_retry#3`.

When creating `tr_retry#3`, the RetriesStatus would contain the status of `tr#0`, `tr_retry#1`, and `tr_retry#2`. So the retry-count is 3, and it equals to the PipelineTask Retries. That's why the example works.

However, the code actually should behave like you say. There is a bug because the maximum number of a TaskRun is Retries+1. The task is declared failure at the second to last retry when number of `RetriesStatus` is == retries. As a result, even though a TaskRun is successful at the last retry, its childTask would never start because it thought the parentTask is already failed.

Reply:
> When creating tr_retry#3, the RetriesStatus would contain the status of tr#0, tr_retry#1, and tr_retry#2. So the retry-count is 3, and it equals to the PipelineTask Retries. That's why the example works.

tr#0 status is captured in `.status.taskRuns[].status`. After tr#0 fails and before creating(updating) taskRun i.e. tr_retry#1, the status is copied to `.status.taskRuns[].status.retriesStatus` and `.status.taskRuns[].status` is cleared which makes `len(retriesStatus)` equal to 1. 

https://github.com/tektoncd/pipeline/blob/2e26eef219ddfabf505041867d4825b47da8cb36/pkg/reconciler/pipelinerun/pipelinerun.go#L656-L657

But the context is applied before copying the status into retriesStatus i.e. `.status.taskRuns[].status` has failure but  `.status.taskRuns[].status.retriesStatus` is empty.

https://github.com/tektoncd/pipeline/blob/2e26eef219ddfabf505041867d4825b47da8cb36/pkg/reconciler/pipelinerun/pipelinerun.go#L456

Reply:
> But the context is applied before copying the status into retriesStatus i.e. `.status.taskRuns[].status` has failure but `.status.taskRuns[].status.retriesStatus` is empty.

I plan to only replace `context.pipelineTask.retries` in this `ApplyContexts`, since it's a const variable. Once the pipelinerun is applied it's gonna not change.

https://github.com/tektoncd/pipeline/blob/9751b951b9f5a35540898db4fb01d1be192dafeb/pkg/reconciler/pipelinerun/pipelinerun.go#L472


As for `context.pipelineTask.retry-count`, the one that will change according to the number of retries, I'm going to replace it here, the `ApplyContexts` for `TaskRun`. This happens after copying the status into retriesStatus. So we don't need to add one, because the length of `.status.taskRuns[].status.retriesStatus` is correct here. But the problem might be whether the name `context.`_**pipelineTask**_`.retry-count` is appropriate or not. 

https://github.com/tektoncd/pipeline/blob/9751b951b9f5a35540898db4fb01d1be192dafeb/pkg/reconciler/taskrun/taskrun.go#L644

To be honest, I planned to replace the both variables in the `ApplyContexts` for `TaskRun`. But it seems impossible to know whether a TaskRun is derived from a PipelineTask unless I add a flag or something similar to TaskRun Spec to identify it.

Reply:
thanks @yaoxiaoqi 👍 

/lgtm

> But the problem might be whether the name context.pipelineTask.retry-count is appropriate or not.

Right, since a taskRun might not belong to any pipeline. Like we discussed earlier, a task will have access to its own retry count so I think its safe to rename the variable to `context.task.retry-count`, similar to `context.task.name` and `context.taskrun.name`.